### ISSUE TITLE
PubSub Subscriber client memory leak

### ISSUE THREAD

**Issue by capt2101akash** [googleapis/python-pubsub#273]
#### Environment details

1. Specify the API at the beginning of the title (for example, "BigQuery: ...") - PubSub 
   General, Core, and Other are also allowed as types
2. OS type and version - linux microservice
3. Python version and virtual environment information: `python --version` Python 3.7

#### Steps to reproduce

 Hi guys, I have created as part of a project a simple subscriber client which get iot data on a real time basis. My subscriber client runs indefinitely and in the callback function I do some transformation in the messages that comes. I have deployed this client as a microservice in GKE, however to utilise GKE more efficiently I was looking into utilising the hpa. I observed that the subscriber client's memory always increases, it never goes down even if the queue is empty. I have tried cancelling thread and closing the subscriber client by putting it in a while loop to release socket resources but in the next iteration, I see the memory has not been freed and it stacks up. I have used memory profiler to detect the memory leak. I know this issue has been discussed in past([#5001](https://github.com/googleapis/google-cloud-python/issues/5001)) but I am utilizing the latest libraries but still facing this. Any help would be really appreciated. 
Providing a sub-sample of code where majority of work happens. I am putting all the messages in a global list so that I can send them for further transformation downstream(If I add the transformation of messages in the callback function the memory consumption increases more rapidly). Adding the GKE dashboard screenshot too for memory consumption and CPU usage. My expectation and the general workflow should ideally be whenever there are spikes in CPU i.e whenever messages are received from queue the memory consumption should increase and after that it should release the memory. 

```Code Sample
def message_ack(message):
    global message_list
    message_list.append(message)
    message.ack()

@profile
def pull_messages():
    future = subscriber.subscribe(subscription_name, message_ack, flow_control=flow_control)
    try:
        future.result(timeout=300)
    except TimeoutError: 
        # logger.error(traceback.print_exc())
        future.cancel()

if __name__ == '__main__':
  flow_control = pubsub_v1.types.FlowControl(max_messages=30)
  subscriber = pubsub_v1.SubscriberClient()
  while 1:
        message_list = []
        pull_messages()
```
![GKE dashboards](https://user-images.githubusercontent.com/16082168/104323627-68bb3580-550c-11eb-8869-a550f64b3b90.PNG)

Thanks!

**Comment by plamut** [comment]
@capt2101akash Thanks for the report.

As a start, could you please share the library versions that your application uses, i.e. the output of `pip freeze`? Also, could you also share a sample code that reproduces the issue if that's possible? How many messages are we generally talking about here? Over what period of time was the memory consumption measured?

The subscriber client has an internal waiting queue of all received messages and only dispatches a subset of them to the user-provided callback (i.e. the flow control), while automatically extending acknowledge deadlines for these messages and such. If these Python instances of these messages are kept alive for some reason, that could explain the ever-reporting memory usage growth as the new messages are streamed from the server - but this is just an educated guess for now.

Any additional info would be appreciated, looking forward to it!

**Comment by capt2101akash** [comment]
> @capt2101akash Thanks for the report.
> 
> As a start, could you please share the library versions that your application uses, i.e. the output of `pip freeze`? Also, could you also share a sample code that reproduces the issue if that's possible? How many messages are we generally talking about here? Over what period of time was the memory consumption measured?
> 
> The subscriber client has an internal waiting queue of all received messages and only dispatches a subset of them to the user-provided callback (i.e. the flow control), while automatically extending acknowledge deadlines for these messages and such. If these Python instances of these messages are kept alive for some reason, that could explain the ever-reporting memory usage growth as the new messages are streamed from the server - but this is just an educated guess for now.
> 
> Any additional info would be appreciated, looking forward to it!

Hey plamut, I was in process of doing that :). Thanks for starting this thread. I have updated my comment with sample code and the charts for memory consumption and CPU usage. I will add the pip freeze output here. In addition to that, I have tried changing the flow controls too to see if that was the case but it was a dead end. Anyways the current setting is 30. The number of messages in consideration will be 200-300K in a day, but it is real time messages so it comes all throughout the day. 
```
apache-beam==2.25.0
avro-python3==1.9.2.1
cachetools==4.1.1
certifi==2020.11.8
cffi==1.14.4
chardet==3.0.4
crcmod==1.7
cx-Oracle==8.0.1
dicttoxml==1.7.4
dill==0.3.1.1
docopt==0.6.2
elasticsearch==7.10.0
elasticsearch-dsl==7.3.0
fastavro==1.2.0
fasteners==0.15
future==0.18.2
google-api-core==1.22.1
google-api-python-client==1.8.4
google-apitools==0.5.31
google-auth==1.23.0
google-auth-httplib2==0.0.4
google-auth-oauthlib==0.4.2
google-cloud-bigquery==2.4.0
google-cloud-bigtable==1.6.0
google-cloud-build==2.0.0
google-cloud-core==1.4.3
google-cloud-datastore==1.15.3
google-cloud-dlp==1.0.0
google-cloud-language==1.3.0
google-cloud-pubsub==1.7.0
google-cloud-spanner==1.19.1
google-cloud-storage==1.33.0
google-cloud-videointelligence==1.16.1
google-cloud-vision==1.0.0
google-crc32c==1.0.0
google-resumable-media==1.1.0
googleapis-common-protos==1.52.0
grpc-google-iam-v1==0.12.3
grpcio==1.33.2
grpcio-gcp==0.2.2
hdfs==2.5.8
httplib2==0.18.1
idna==2.10
kafka-python==2.0.2
libcst==0.3.14
memory-profiler==0.58.0
mock==2.0.0
monotonic==1.5
mypy-extensions==0.4.3
mysql-connector-python==8.0.22
mysqlclient==2.0.1
numpy==1.19.4
oauth2client==4.1.3
oauthlib==3.1.0
pandas==1.1.4
pandas-gbq==0.14.1
pbr==5.5.1
proto-plus==1.11.0
protobuf==3.14.0
psutil==5.8.0
pyarrow==0.17.1
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycparser==2.20
pydata-google-auth==1.1.0
pydot==1.4.1
pymongo==3.11.1
pyparsing==2.4.7
python-dateutil==2.8.1
pytz==2020.4
PyYAML==5.3.1
redis==3.5.3
requests==2.25.0
requests-oauthlib==1.3.0
rsa==4.6
six==1.15.0
SQLAlchemy==1.3.20
typing-extensions==3.7.4.3
typing-inspect==0.6.0
uritemplate==3.0.1
urllib3==1.26.2
xmltodict==0.12.0
```
Bear with the large list of packages.

**Comment by capt2101akash** [comment]
Hey @plamut I hope you got some time to see the details that I have provided. Any inputs will be really helpful. Looking forward to it.

**Comment by plamut** [comment]
@capt2101akash I did do a quick check of the sample code and it seems like a straightforward streaming pull usage.

The messages being accumulated in the global `message_list` should be discarded on each iteration when the reference is changed to the new empty list, but there can always be some internal structures in the client that are kept alive and causing the memory growth. It is also possible, however, that all this is normal behavior, since Python does not necessarily release all the allocated memory back to the OS (e.g. due to keeping some of the already allocated memory for future use).

I hope I can take a closer look some time next week, but until then and if the time allows you, you can perhaps experiment with the following:
- Run an explicit `import gc; gc.collect()` after re-assigning `message_list` to a new list to clean up some unused objects.
- Instead of using a common subscriber instance, create a new subscriber in `pull_messages()` and call the [close()](https://github.com/googleapis/python-pubsub/blob/542d79d7c5fb7403016150ba477485756cd4097b/google/cloud/pubsub_v1/subscriber/client.py#L223-L231) method when done with it (or use it with a context manager).
  If the cleanup triggered by `future.cancel()` does not correctly clean all underlying resources, this might help.
- Run `pull_messages()` in a subprocess in each iteration. When a child process terminates, the OS will reclaim all its memory.
  This might require changes to any logic in the main process that collects the results (if it's not process-and-forget), but it's an often used technique when performing memory-intensive tasks.

**Comment by capt2101akash** [comment]
@plamut thanks for your suggestions. However, I have tried all of the above that you mentioned but to no avail. I will still experiment a bit more on the multiprocessing part though and will let you know my findings. In the meantime I will be looking forward to get more insights from you sometime next week.

**Comment by plamut** [comment]
Multiprocessing trick should almost certainly work, but of course if there's a leak, we should find and fix it. Could be tricky, though. :)

**Comment by capt2101akash** [comment]
> Multiprocessing trick should almost certainly work, but of course if there's a leak, we should find and fix it. Could be tricky, though. :)
> 
> **Edit:** Oh, you tried that already? That's somewhat surprising. Could there be anything else outside of the provided sample that is consuming memory, or is the sample fully self-contained?

The sample is self contained and even if we just do plain subscriber client the memory is increasing. Multiprocessing way is somewhat tricky though as we need to maintain a global list. I am looking into a workaround to do that. In any case, we should really see what's causing this gradual increase so that we can fix that. Can you once try reproducing the same and see if it behaves the same way for you too or not?

**Comment by capt2101akash** [comment]
@plamut Update:- As expected multiprocessing pathway is also a dead end, as grpc doesn't work with multiprocessing. Reference here #3890 . I hope after you do some research we get to know more about this issue.

**Comment by plamut** [comment]
@capt2101akash Thanks for the research, although the linked issue seems a bit different - the user-provided callbacks are, by default, invoked using an internal thread pool executor, while the linked issue is about using a process pool executor for dispatching the messages (which doesn't seem to work due to the underlying `grpc` dependency).

What I meant was running the entire streaming pull and message processing in a child process (including the client instantiation), so that any memory used (or leaked) would be automatically reclaimed by the OS upon the child process termination. Something like the following  (note: untested):

```py
from multiprocessing import Process

...

def pull_messages():
    flow_control = pubsub_v1.types.FlowControl(max_messages=30)
    subscriber = pubsub_v1.SubscriberClient()
    future = subscriber.subscribe(subscription_name, message_ack, flow_control=flow_control)
    try:
        future.result(timeout=300)
    except TimeoutError: 
        future.cancel()


if __name__ == '__main__':
    while True:
        p = Process(target=pull_messages)
        p.start()
        print(f"Started a new child process {p.pid}")
        p.join()

```

It would require some adjustments to get any message processing results back to the main process, if needed, but I hope you get the idea.

---

BTW, a random observation - I see that `FlowControl.max_messages` is set to 30, but the [default thread pool executor](https://github.com/googleapis/python-pubsub/blob/542d79d7c5fb7403016150ba477485756cd4097b/google/cloud/pubsub_v1/subscriber/scheduler.py#L67-L73) uses 10 worker threads. If the message callback does any non-trivial CPU work, only a maximum of 10 messages will be dispatched at a time, because the rest would have to wait for the worker threads to free up.

**Comment by capt2101akash** [comment]
@plamut Yeah I got what you are saying. The same piece of code that you sent, I tried that out yesterday. The only concern here is as you said get the message processing results back to the main process. And this is where I felt the referenced issue was coming into place. Sorry if I didn't clearly specify it earlier, but here is what I meant. 
- Once we send the pull_messages in a separate process, we need a global variable to store our messages in a list. To do that I used multiprocessing.Manager().list() 
- Once we call pull_messages, there is a callback function as you have written *message_ack*. This callback function should ideally append the messages in the global list created in the first step. 
- The problem comes here, once we call the list.append(messages) it fails with error `TypeError: can't pickle _thread.lock objects`. As per my research this is related to multiprocessing variables not being able to used by threads. Correct me if I am wrong and this is where I was referring to the previous issue. My bad that I didn't properly state.

```
import multiprocessing

def message_ack(message):
    # global message_list
    message_list.append(message) #Error occurs here 
    message.ack()

def pull_messages(message_list):
    future = subscriber.subscribe(subscription_name, message_ack, flow_control=flow_control)
    try:
        future.result(timeout=30)
    except TimeoutError: 
        # logger.error(traceback.print_exc())
        future.cancel()

if __name__ == '__main__':
  ...
  flow_control = pubsub_v1.types.FlowControl(max_messages=5)
  manager = multiprocessing.Manager()
  while 1:
      message_list = manager.list()
      subscriber = pubsub_v1.SubscriberClient()
      # message_list = list()
      p1 = multiprocessing.Process(target=pull_messages, args=(message_list,))
      p1.start()
      p1.join()
      subscriber.close()

```

**NOTE**- As you can see I have changed the flow control to get only 5 messages and this is also not resulting in less memory consumption. The memory gradually grows.

I hope this clears out the misunderstanding. If you have any inputs on how we can mitigate the error related to can't pickle thread.lock objects it will be helpful. I have added the line where we get that error. 
Hoping to hear back from you soon.

**Comment by plamut** [comment]
@capt2101akash Thanks for the clarification, it indeed seems we're on the same page.

From the top of my head - the messages dispatched to the call are [Message](https://github.com/googleapis/python-pubsub/blob/542d79d7c5fb7403016150ba477485756cd4097b/google/cloud/pubsub_v1/subscriber/message.py#L53-L74) instances that wrap the actual messages received from the server.

These instances contain, among other things, a reference to a queue used for asynchronously dispatching requests such as ACK, NACK, and such, but the queue itself cannot be pickled:
```py
>>> import queue
>>> import pickle
>>> pickle.dumps(queue.Queue())
TypeError                                 Traceback (most recent call last)
<ipython-input-43-96dc0f23ad59> in <module>
----> 1 pickle.dumps(queue.Queue())

TypeError: cannot pickle '_thread.lock' object
```

Since all the `message_ack()` callback does before acknowledging a message is appending it to a list, there's probably no need to add the entire `Message` instance, but instead just the message contents? Something like the following:
```py
def message_ack(message):
    message_list.append(message.data)
    message.ack()
```

Would that be sufficient for the post-processing the received message data accumulated in the list?

**Comment by capt2101akash** [comment]
@plamut I missed that part, using message.data does indeed work with multiprocessing(I missed that part). However, when I am trying to do my processing inside the callback(other than just message.ack()) it fails with segmentation fault. In addition to that, it behaves very sporadically and fails with segmentation fault quite frequently even though if we just do message.ack() in callback function. I wonder if it is related to the issue I mentioned, but I am again feeling at the dead end as nothing seems to working out, if at all we find out the specifics into why the subscriber thread is not releasing the memory. If you have encountered the segmentation faults in multiprocessing, do let me know what you did for mitigating it. 
Looking forward for your reply.

**Comment by plamut** [comment]
@capt2101akash 

> But in the post processing if something goes bad like the issue mentioned above, the messages get lost and we won't be able to get them back. So ideally, I we should do all the processing first and then acknowledge the messages. If that makes sense.

That makes perfect sense. The backend will not try to re-send a message after it has received an ACK (within the message's acknowledge deadline). Applications should thus send an ACK only after they have processed a message. If processing means more than merely appending a message to a list (and then perhaps storing it somewhere), `ack()` should be postponed until then.

In principle, one could append a message _and_ its ACK ID to the list and then manually send an acknowledge requests for the processed messages, but that kind of beats the purpose of the client that relieves the programmer from tasks such as retries, automatic lease management, etc., thus I wouldn't go down that route.

I can't comment on the segfaults in the callback without the code, but if they frequently happen even when just calling `message.ack()`, then it indeed appears multiprocessing is a dead end, unfortunately. I'll try to reproduce segfaults, too, and report back.

---

Random though - is there a specific reason for running the streaming pull for a fixed amount of time, cancelling it, and then starting it again? Was that just for the purpose of profiling?

Would your use case allow for moving this loop out of the subscriber script itself? By that I mean running it for a limited amount of time, processing the message stream, and then letting it terminate. After that (or even a parallel) another subscriber is fired up to continue the processing. Kind of the same idea as before, but without a child process.

(I do realize this is not ideal, but mostly trying to come up with a reasonable workaround for the time being)

**Comment by capt2101akash** [comment]
@plamut per your point regarding running the streaming pull for a fixed amount of time and cancelling it has two uses:-
- As you mentioned, I did it for profiling. 
- Secondly, to release resources in a continuous running thread was my first thing to cross off as it will be continuously running and the behavior we are facing here would have been there too. I chose this way to make sure that once my future gets cancelled, I release all resources to the os.

To your suggestion, are you suggesting that we should have a cronjob kinda thing (I am also thinking that). If that is the case, then yes it will be our last resort (not ideal for real time cases). 
To help you in reproducing in the seg fault issue, try to do some complex computation in the callback for the processing of messages, then send an ack. This will help to get to that issue (in my case I was querying a sql database by using a key from json as parameter).

**Comment by plamut** [comment]
After testing a good number of combinations - Python version, `grpcio` version, test script flavour - it appears that there indeed exists a memory leak, but to make it more apparent, the number of messages posted to a topic should be high.

I used the initial code sample as a base, ran the streaming pull in the `pull_messages()` for 10 iterations lasting ~30 seconds each (the timeout passed to the streaming pull future), and parallel posted 50.000 messages to the topic in quick succession. Each message was a random ASCII string of ~4100 characters.

![memusage_base_example_py38_grpcio1 34 1](https://user-images.githubusercontent.com/1235663/105176181-32da0a80-5b25-11eb-9dd9-cf98af90dabb.png)

Creating a subscriber inside the `pull_messages()` each time and closing it at the end (used it as a context manager) did not really help, the memory consumption pattern looked similar:
![memusage_client_in_function_py38_grpcio1 34 1](https://user-images.githubusercontent.com/1235663/105175655-8a2bab00-5b24-11eb-8fa2-bee0ed0e5766.png)

However, if running `pull_messages()` in a a child process, the memory consumption was stable:
![memusage_client_in_subprocess_py3 8_grpcio1 34 1](https://user-images.githubusercontent.com/1235663/105176724-e3e0a500-5b25-11eb-9a60-cc2b945834d0.png)

I was also not able to reproduce any segfaults despite doing non-trivial work in the message callback (queried a simple database), although absence of evidence is not an evidence of absence, of course. But if I understand correctly, that should be easily reproducible, @capt2101akash ?

It might be worth noting that I created, ran, and closed a subscriber instance _inside_ the `pull_messages()` function, i.e. _after_ the fork.

**Comment by capt2101akash** [comment]
@plamut I can try creating subscriber in the sub process and revert back to you my findings. However, I think we can 100% agree to the point that there is a memory leak that has to be addressed, as running streaming pull in subprocess is just a workaround that we are using to overcome this issue. I would be really glad if we can pin point the module/functionality which is causing this leak and fix this. 
In the meantime I will give me results back to you(I don't recall if I tried that or not) but I will try it again and let you know.
Thanks for your time man! Really appreciate the help!

**Comment by plamut** [comment]
@capt2101akash The leak should definitely be fixed, indeed, but it's always good to have a reasonable workaround at hand - fingers crossed that it will work for you! :crossed_fingers:

**Comment by capt2101akash** [comment]
Just to give you a good news. It did work 😊. I somehow never tried this approach 😅 my bad.
I will be monitoring my code for a couple of days to see how it is performing or are there any issues that come up down the line. However, let's see how we can fix the memory leak. 👍 Thanks again

**Comment by plamut** [comment]
I would like to give an update on the most recent findings.

I tried running the streaming pull in a single process and for longer periods of time (15+ minutes as opposed to only a few minutes). I tried both with a plain callback logic, i.e. only calling `message.ack` and with the version that appends the message to a global list (the list is re-created on each streaming pull iteration, each iteration took ~60 seconds).

In all runs I made sure that the subscription always had enough messages at hand to keep the subscriber as busy as possible (so that flow control had to kick in).

### 15 x 60 seconds, no append
<details>
    <summary>Expand to view graphs</summary>

![run_15_minutes_no_append](https://user-images.githubusercontent.com/1235663/106438978-7a428e00-6477-11eb-8367-d53e7e7c9e21.png)

![run_15_minutes_no_append_2](https://user-images.githubusercontent.com/1235663/106438992-7dd61500-6477-11eb-9a60-fb78926773a6.png)
</details>

Initially, the memory consumption slowly increases, but eventually flattens out. This hints that any internal streaming pull operations probably do not leak memory.


### 15 x 60 seconds, appending to a list
<details>
    <summary>Expand to view graphs</summary>

![run_15_minutes_append](https://user-images.githubusercontent.com/1235663/106439818-6b101000-6478-11eb-94ea-c96367cc0e4c.png)

![run_15_minutes_append_2](https://user-images.githubusercontent.com/1235663/106440041-a9a5ca80-6478-11eb-8231-23edbb88769f.png)

![run_15_minutes_append_3](https://user-images.githubusercontent.com/1235663/106440062-af9bab80-6478-11eb-923a-43ce15e59ec9.png)

</details>

Since received messages are temporarily stored in a temporary global list, the memory consumption increases faster. However, the growth is not linear - there are points when some of the memory is released back to the OS. It appears that after the initial increase, the memory consumption actually remains bounded.

If we only looked at the first few minutes, it might indeed seem that there is a memory leak, but monitoring the process over a longer time span does not support that claim.


### 30 x 60 seconds, appending to a list

<details>
    <summary>Expand to view graphs</summary>

![run_30_minutes_append](https://user-images.githubusercontent.com/1235663/106441111-f047f480-6479-11eb-8c0e-fa5d43dbf891.png)

</details>

Since the second 15-minute graph was a bit inconclusive and somewhat differed from the other two, I did yet another run for full 30 minutes. The graph confirmed that the memory consumption is, in fact, bounded.

The most likely reason for seemingly "leaking" memory at the beginning is CPython's own memory management. When an object such as a list of messages goes out of scope, the memory is not immediately released back to the OS. Instead, the interpreter keeps it for future use, as it assumes that more objects will be created, and re-using the memory from an already allocated memory pool is faster than constantly freeing the memory to the OS and then allocating it back.

---

@capt2101akash
The big factor in memory consumption is whether the messages are stored in a list or not, but even if they are, the graphs suggest that the memory does **not** seem to be leaking, at least not by the subscriber client itself. Could it be that the memory consumption increases due to the application storing the results of the processing somewhere without regularly releasing these results?

I tried to find any zombie Message (or other) instances lurking in the dark and forgotten corners of memory, but couldn't really find any, they seem to be garbage collected just fine. I also checked the streaming pull code for any queues, lists, etc. that could grow indefinitely. 

Initially, one possible candidate was the histogram that tracks how long does it take to acknowledge messages on average, but the the histogram size is bounded. The [add()](https://github.com/googleapis/python-pubsub/blob/8751bcc5eb782df55769b48253629a3bde3d4661/google/cloud/pubsub_v1/subscriber/_protocol/histogram.py#L108-L125) method maps all ACK times onto the interval `[10, 600]`, and the ACK times tracked are [rounded](https://github.com/googleapis/python-pubsub/blob/8751bcc5eb782df55769b48253629a3bde3d4661/google/cloud/pubsub_v1/subscriber/message.py#L219-L227) to integer values, meaning that the histogram stats are bounded - they can only contain a maximum of 591 distinct values.

**Comment by capt2101akash** [comment]
@plamut I have also observed this when the frequency of messages coming in the topic is not real time or to be precise not that frequent. However, if you have seen my previous comments wherein I have posted the snippet, where I also tried just acknowledging the messages without any append and with a continuous stream of messages flowing in at the rate of 100 messages/sec, the graph shows gradual increase. However, as you have mentioned it does release some memory in between but overall it increase in a period of a day or two. 
I remember in the previous [comment](https://github.com/googleapis/python-pubsub/issues/273#issuecomment-763591482) too you have mentioned that you saw a leak with just using the initial base file. 
In any case however, my point is even if I don't append anything to the global list, there is a leak.
I would say try with large number of messages coming every single minute or so and then process them with simple basic code to just acknowledge the message as you have done before. I am pretty sure that you will see a leak happening as you did before.

Just a quick update to where I currently stand now for my application so that others if they are facing any issues can correlate to this - I ran my subscription client in a multiprocessing environment, which after completion releases it's memory back to OS and I get what I required. Just a thing to keep in mind though is that one might have to adjust their processing logic in order to utilise this solution effectively.

**Comment by plamut** [comment]
@capt2101akash Thanks for the reply, and I'm glad that the multiprocessing workaround is working well.

> However, if you have seen my previous comments wherein I have posted the snippet, where I also tried just acknowledging the messages without any append and with a continuous stream of messages flowing in at the rate of 100 messages/sec, the graph shows gradual increase. However, as you have mentioned it does release some memory in between but overall it increase in a period of a day or two.

FWIW, in the test that I conducted the subscriber running on my laptop received and processed (i.e. appended to a list) somewhere around 6500 messages / minute, which is comparable to the load in your test (the message stream was constantly high). I did not run it for the entire day, though.

> I remember in the previous [comment](https://github.com/googleapis/python-pubsub/issues/273#issuecomment-763591482) too you have mentioned that you saw a leak with just using the initial base file.

Yeah, that's what it looked like, although that test only ran for 5 minutes or so. Running the same test longer later revealed that the memory consumption actually stabilizes eventually.

Based on your reports, it would probably still make sense to set up a small cloud app with a busy publisher and a simple ACK-only subscriber, and let it run for at least 24 hours. I'll try to do that when I get back to this.

**Comment by lidizheng** [comment]
Hi @plamut, it might help generate a memory profile using [memory_profiler](https://github.com/pythonprofilers/memory_profiler). It describes all living objects in Python space. If we do observe a large number of some class, it indicate the leak happen in Python space. Otherwise, it might caused by C/C++ objects leaking. Here is one memory leak fix in gRPC: https://github.com/grpc/grpc/issues/22123#issuecomment-623626840.

**Comment by plamut** [comment]
@lidizheng Thanks, that's precisely the tool I used, but I didn't find any weird orphaned Python objects, and the time graphs so far could not support the claim that there is a leak (at least not when measuring locally over 30 minutes).

**Comment by manuelfedele** [comment]
Hello everyone.

I'm investigating this too, because I'm facing a serious issue with memory intensive callback using subscriber.

You can find the repo with code, .dat and .png files here https://github.com/manuelfedele/google-pubsbub-memory-leak

**Details**:
- Ubuntu 20.04
- Python 3.8

Only 3 requirements:
- google-cloud-pubsub==2.4.1
- matplotlib==3.4.1
- memory-profiler==0.58.0

### Create a simple "memory intensive" callback and run it as pure python callback **4** times
```python
from memory_profiler import profile


@profile
def callback():
    # Memory intensive operation
    x = [n for n in range(int(1e5))]


if __name__ == '__main__':
    callback()
    callback()
    callback()
    callback()
 ```
![pure_function](https://user-images.githubusercontent.com/22051383/114271629-9b44d380-9a12-11eb-88d2-5cf3f05974ca.png)

Everything looks fine here to me.
___
### Use the same "memory intensive" callback as pubsub callback and send **4 messages** 
```python
import os
from google.cloud import pubsub_v1
from memory_profiler import profile

os.environ.setdefault('GOOGLE_APPLICATION_CREDENTIALS', 'path/to/key.json')
os.environ.setdefault('GOOGLE_CLOUD_PROJECT', 'project-id')
os.environ.setdefault('MY_TOPIC_NAME', 'test_memory_leak-sub')

topic_name = 'projects/{project_id}/topics/{topic}'.format(
    project_id=os.getenv('GOOGLE_CLOUD_PROJECT'),
    topic=os.getenv('MY_TOPIC_NAME'),  # Set this to something appropriate.
)

subscription_name = 'projects/{project_id}/subscriptions/{sub}'.format(
    project_id=os.getenv('GOOGLE_CLOUD_PROJECT'),
    sub=os.getenv('MY_TOPIC_NAME'),  # Set this to something appropriate.
)


@profile
def callback(message):
    # Memory intensive operation
    x = [n for n in range(int(1e5))]
    message.ack()
    print("ack")


with pubsub_v1.SubscriberClient() as subscriber:
    future = subscriber.subscribe(subscription_name, callback)
    print("Starting subscriber")
    try:
        print("Listening...")
        future.result(timeout=30)
    except KeyboardInterrupt:
        future.cancel()
```
![subscriber_intensive](https://user-images.githubusercontent.com/22051383/114271713-1a3a0c00-9a13-11eb-8eee-dca1642014d2.png)

Here we have a serious memory leak. **It is evident because of the "huge" callback memory usage (?!)**
___
#### Remove the memory intensive part from the callback, use it as subscriber callback and send **4 messages** 
```python
import os
from google.cloud import pubsub_v1
from memory_profiler import profile

os.environ.setdefault('GOOGLE_APPLICATION_CREDENTIALS', 'path/to/key.json')
os.environ.setdefault('GOOGLE_CLOUD_PROJECT', 'project-id')
os.environ.setdefault('MY_TOPIC_NAME', 'test_memory_leak-sub')

topic_name = 'projects/{project_id}/topics/{topic}'.format(
    project_id=os.getenv('GOOGLE_CLOUD_PROJECT'),
    topic=os.getenv('MY_TOPIC_NAME'),  # Set this to something appropriate.
)

subscription_name = 'projects/{project_id}/subscriptions/{sub}'.format(
    project_id=os.getenv('GOOGLE_CLOUD_PROJECT'),
    sub=os.getenv('MY_TOPIC_NAME'),  # Set this to something appropriate.
)


@profile
def callback(message):
    # Memory intensive operation
    # x = [n for n in range(int(1e5))]
    message.ack()
    print("ack")


with pubsub_v1.SubscriberClient() as subscriber:
    future = subscriber.subscribe(subscription_name, callback)
    print("Starting subscriber")
    try:
        print("Listening...")
        future.result(timeout=30)
    except KeyboardInterrupt:
        future.cancel()
```
![subscriber_ack_only](https://user-images.githubusercontent.com/22051383/114271739-4786ba00-9a13-11eb-9e4c-b563ea01f0ea.png)

Here we can't notice a memory leak, because we sent only 4 messages, but it is there.  **It is not evident because of the low callback memory usage (?!)**
___

## Conclusion: The subscriber is leaking memory and the amount of memory leaked is directly proportional to the underlying callback memory usage.

#### **It looks like it's "caching" the callback, but why? This gets even worse if you run this test with 50 messages.**

Am I doing something wrong? What am I missing?
I'm using the subprocess workaround at the moment on my containers, but I don't really like it.

_The subscriber instantiation is copy pasted from the README here on github._

Hope this helps.

**Comment by manuelfedele** [comment]
Maybe that the problem may reside here https://github.com/googleapis/python-pubsub/blob/9f451a1fe0ac5cb2fb13d72b0436e0b521a4fecb/google/cloud/pubsub_v1/subscriber/scheduler.py#L117
and could be related to a bug in ThreadPoolExecutor https://bugs.python.org/issue38430

It should be awaited?

**Comment by plamut** [comment]
@manuelfedele Thanks for the research on your part, and the results you posted.

It seems weird that temporarily allocating a list in the callback would cause the leak. When the callback returns, its stack frame is gone, and so are the references to local variables in the callback - unless something else keeps a reference to that frame. This is the result I got:

![reproduce_report_no_import](https://user-images.githubusercontent.com/1235663/114377958-83db2700-9b87-11eb-88b1-25f094f53f39.png)

I published 30 messages to the topic and added a brief `time.sleep(1)` after allocating the list in the callback, but before acknowledging the received message.  There were 10 callbacks executed in parallel, because that's the default size of the internal thread pool.

The memory usage did not grow linearly with each callback, and actually dropped back before stabilizing when no more messages were arriving anymore.

It's worth mentioning that I did **not** import the `profile` decorator explicitly. However, if I added the following import at the top:
```py
from memory_profiler import profile
```
...and ran a profiling session as before:
```
$ mprof run --include-children my/test/script.py
```
...then I got the same graph as yours.

![reproduce_report_with_import](https://user-images.githubusercontent.com/1235663/114379372-07494800-9b89-11eb-9116-dd8f48020a80.png)

In the latter case, I also noticed that the messages were not being acknowledged, there were no `"ack"` strings printed out, and the script did not exit normally after timeout, I had to stop it with a keyboard interrupt.

It seems that explicitly importing the `profile` decorator causes some issues. Can you also try profiling without the import and report if your observations are consistent with mine?

P.S.: Even though this particular case appears to be some sort of a profiler issue, the leak is nevertheless present as previously demonstrated by a long-running app and observing its memory consumption growing over time.

**Comment by manuelfedele** [comment]
Hello @plamut, thank you for your check

I'll test it asap, but I can tell you for sure that the second graph (linear increment) is the one that GCP shows in memory explorer when I run that very container in production.

Anyway I tried to implement my own scheduler in several ways.
I'm not sure about this, but I noticed that if I create a scheduler that inherits from [ThreadScheduler ](https://github.com/googleapis/python-pubsub/blob/9f451a1fe0ac5cb2fb13d72b0436e0b521a4fecb/google/cloud/pubsub_v1/subscriber/scheduler.py#L82) and make some changes to the `schedule` method, the leak itself changes.

For example, a nonsense implementation like this below:

```python
class MyScheduler(ThreadScheduler):

    def __init__(self, executor=None):
        super(MyScheduler).__init__(executor=executor)
    
    def schedule(self, callback, *args, **kwargs):
        """Schedule the callback to be called asynchronously in a thread pool.
        Args:
            callback (Callable): The function to call.
            args: Positional arguments passed to the function.
            kwargs: Key-word arguments passed to the function.
        Returns:
            None
        """
        # self._executor.submit(callback, *args, **kwargs)
        # async to sync naive refactor to test memory leakage
        callback(*args, **kwargs)
```

Does not leak memory, but why? The queues are filled the same as before and I think that everything else under the hood works the same way (it should be even worse; all jobs should be "pending" in the _ThreadScheduler_'s queue if they are not submitted to the executor, right?)

Can the leak be somehow related to this?
https://stackoverflow.com/questions/44403132/close-python-thread-to-prevent-memory-leak

(check this too: https://bugs.python.org/issue38430)

**Comment by plamut** [comment]
> I'll test it asap, but I can tell you for sure that the second graph (linear increment) is the one that GCP shows in memory explorer when I run that very container in production.

I can confirm this, yes.

I originally assumed that the underlying `ThreadPoolExecutor` can be excluded, because it's in a standard library and used by so many people that _surely_ one would have noticed the leak already if there was one. But perhaps we should re-think that and see what's going on - either in the thread pool or in the way we interact with it.

I appreciate the research you have done on your own so far!

**Comment by manuelfedele** [comment]
That's correct.

Wrapping and profiling in main and set list size to 1e7 shows a behavior like the one you posted. The new graph is better, but setting the list size to 1e7 is just hiding the leak.

I did this test to show it better.
1 - Wrapped and profiled main() as you said
2 - Set the list size to 1e6 instead of 1e7 to avoid to hide the leak
3 - Sent 40 messages in batches (10 messages/batch, 5 seconds sleep between them)

This is what I got:
![profile_main_1e6](https://user-images.githubusercontent.com/22051383/115114322-f0429580-9f8e-11eb-8302-1bf705377982.png)

The leak is still there. :(

**Comment by plamut** [comment]
> The new graph is better, but setting the list size to 1e7 is just hiding the leak.

Maybe I wasn't clear enough, that test was mainly to see that the leak probably cannot be attributed to temporarily allocating the memory in the callback (which would be quite unexpected, frankly). That memory does get released, while the leak itself happens elsewhere (C level is suspected).

FWIW, I remember that the leak becomes more apparent if more/larger messages are received, it shows up on the graph more quickly.

**Comment by manuelfedele** [comment]
Yep, you were clear, but that's not correct. :) 

This is my container, with 1Gb RAM.
Crashed after 5 messages with intensive callback on GCP (sequential messaging, non contemporary)
![image](https://user-images.githubusercontent.com/22051383/117287507-b5839d00-ae6a-11eb-9f2a-e4b8aad336b5.png)

This is another container, with a non intensive callback. Leaking but not crashing with > 100 messages.
![image](https://user-images.githubusercontent.com/22051383/117287772-fed3ec80-ae6a-11eb-923e-6ef6d103d84b.png)


I think that is clear that the memory allocated by the callback is not released.
This can't be a wrong, because the first container crashed. The second one didn't.

**Comment by plamut** [comment]
@manuelfedele A [fix](https://github.com/googleapis/python-pubsub/pull/406) for the discovered publisher memory leak inspired me to have another quick look at this. That fix would not help here, unfortunately, because the relevant threads are already daemonic, but I did observe that a leak exists regardless whether the callback is executed through the executor, or by directly calling it.

The previous theory that the leak might be caused by the executor ([comment](https://github.com/googleapis/python-pubsub/issues/273#issuecomment-818497905), [comment 2](https://github.com/googleapis/python-pubsub/issues/273#issuecomment-818507621)) is probably not correct after all. I also couldn't find any executor-related memory issues in the CPython's bugtracker, except [the one](https://bugs.python.org/issue38430) that you already linked, but that one is about not awaiting a future returned when submitting to an executor.

TL; DR - indeed not all memory allocated by the callback seems to be released, it's probably just not that temporary list but something else.

FWIW, I managed to reproduce this with `@profile` applied to the `main()` function, not the callback itself, thus the previous concern about the profiler interfering with the streaming pull too much can likely be discarded. I'll try to re-visit this again in a few days, there are a few additional experiments I'd like to run to (dis)prove possible culprits.

**Comment by capt2101akash** [comment]
Hey guys, thanks for continuing this thread, I agree with the theory suggested by @manuelfedele with high frequency of messages in pubsub increases the leak. And @plamut as we discussed earlier I hope we try to solve this as it gets apparent when we do a real time streaming in k8s. 
The way that I handled it is a way around which is incorporating it in multiprocessing, this resulted in desired output but it should not be done in this way. 
Hoping to get more ideas here soon. Will do my part of research too and post here.

**Comment by plamut** [comment]
FWIW, the reason why this was marked as blocked is that the feedback time can be long. Even running and profiling a sample app locally for half an hour might not always be enough to definitely conclude whether a leak is present or is it just a short-term noise (some of the memory usage [graphs](https://github.com/googleapis/python-pubsub/issues/273#issuecomment-770738240) produced by the profiler turned out to be misleading).

**Comment by synackSA** [comment]
Just wondering what the current status of this is. We're also experiencing a memory leak issue, this seems to persist even if I fire up the box via docker-compose and use the pubsub emulator on my localhost. If I monitor the box resources with `top`, the overall memory usage goes up, but the individual threads for Python seem to stay the same.

Some Examples:

After processing 10k+ messages, the resources for the box and processes looks like the following:
<img width="781" alt="Screen Shot 2021-10-06 at 11 52 28 AM" src="https://user-images.githubusercontent.com/5345876/136252206-80f189f5-4997-4617-9a0b-077b2b19b132.png">

We then processed 5k+ messages and the memory for each process stayed the same, but the overall box memory consumption had increased (from 19.4% to 19.7%):
<img width="785" alt="Screen Shot 2021-10-06 at 1 07 44 PM" src="https://user-images.githubusercontent.com/5345876/136253146-23f19035-7bc1-4313-b470-ef5e85d2aeab.png">

We then processed another 7.5k messages. In this case, we see a small increase in memory PID 23 (105136 -> 105532), but nothing that would account for a increase from 19.7% to 20.1%
<img width="784" alt="Screen Shot 2021-10-06 at 1 33 55 PM" src="https://user-images.githubusercontent.com/5345876/136254958-82a424ba-5824-48c8-b561-fb445f104f8e.png">

I'm currently in the process of trying to use mprof to profile and graph too, so we'll see what results we get from that.

**Comment by synackSA** [comment]
Just want to follow up on this and see if there's been any movement on this? We've implemented the design where we spawn another process, which dies after a certain amount of time, get's cleaned up and then spawn a new one. However, I'd like to keep tabs on this issue to see what's happening.

**Comment by acocuzzo** [comment]
Thanks very much @gberth

**Comment by blacar** [comment]
BTW I dont get it ... this is a memory leak on an official Google library that has been alive for more than 1 year (?!)
Is python a pariah on GCP????

**Comment by gberth** [comment]
Without any evidence, it is my opinion/experience that it also comes to the use of Bigtable from python. So I suspect it has something to do with more low-level code than pubsub itself. I agree that the absence  of priority and development is frustrating. It consumes a lot of time that could/should have been used on business development

**Comment by manuelfedele** [comment]
@blacar 

Sadly, we had to change cloud provider because of this.
It was unsustainable (crash every 10 messages in our case) and the multiprocessing trick was slowing down our pipeline.
It is reproducible. But in backlog since more than one year now. So we could not trust the ecosystem.

Very sad.

**Comment by manuelfedele** [comment]
Well that's not strange. What we noticed is that the whole memory taken by the process is not released. In our case, we had 1024Mb of ram. Every request, took about 200Mb of ram. Even if sent sequentially (waiting 10 seconds between the messages), after 5-6 messages the container was crashing for OOM error.

Maybe your operation is not memory intensive.

Sadly, restarting every 5 or 6 messages was not acceptable for us.

**Comment by gberth** [comment]
Anything at all going on with this?

**Comment by gberth** [comment]
A bit frustrating that this never gets priority. And why is it blocked?

One observation, it seems that the memory usage increases faster proportional with the (time or occurrence of) reaching of max-messages.  

rgds 
Gunnar

**Comment by manuelfedele** [comment]
Could this be related to https://github.com/grpc/grpc/issues/28747 ?

**Comment by acocuzzo** [comment]
Following up with grpc folks to see when this fix will be in a grpc release (currently the change is checked in, but not included in the latest release). As soon as it is in a release, we will update the client library.

**Comment by acocuzzo** [comment]
According to grpc folks, this change will be included in grpc release: v1.51.0, which will be out in approx 1-2 weeks.

**Comment by acocuzzo** [comment]
The pre-release v1.15.0-pre1 contains the fix:
https://github.com/grpc/grpc/releases/tag/v1.51.0-pre1

**Comment by acocuzzo** [comment]
The release fix is out: https://github.com/grpc/grpc/releases/tag/v1.51.0
